src/clp_ffi_js/ir/ClpIrV1Decoder.hpp

```diff
@@ -0,0 +1,104 @@
+     * includes the encoding type.
+     * - Passing those to the ClpIrV1Decoder() for their life cycles management.
+     *
+     * @param data_array JS Uint8Array which contains the Zstd-compressed CLP IR V1 bytes.
+     * @return the created instance.
+     * @throw DecodingException if any error occurs.
+     */
+    [[nodiscard]] static auto create(emscripten::val const& data_array) -> ClpIrV1Decoder;
+
+    // Destructor
+    ~ClpIrV1Decoder() = default;
+
+    // Explicitly disable copy constructor/assignment
+    ClpIrV1Decoder(ClpIrV1Decoder const&) = delete;
+    auto operator=(ClpIrV1Decoder const&) -> ClpIrV1Decoder& = delete;
+
+    // Define default move constructor
+    ClpIrV1Decoder(ClpIrV1Decoder&&) = default;
+    // Delete assignment operator due to Emscripten (derived from LLVM)
+    // being unable to resolve the move constructor for clp::ir::LogEventDeserializer
+    auto operator=(ClpIrV1Decoder&&) -> ClpIrV1Decoder& = delete;
+
+    /**
+     * Retrieves an estimated number of log events.
+     *
+     * @return The number of log events deserialized if `build_idx()` has been called and returned
+     * successfully.
+     * @return 0 (cFullRangeEndIdx) to indicate there is no event has been deserialized yet by
+     * `build_idx()`.
+     */
+    [[nodiscard]] auto get_estimated_num_events() const -> size_t;
+
+    /**
+     * Deserializes log events in the range `[beginIdx, endIdx)` and buffers the results. Currently,
+     * only full range build is supported, and the buffered bytes of the stream will be released on
+     * success.
+     *
+     * @param beginIdx
+     * @param endIdx
+     * @return Count of the successfully deserialized ("valid") log events and count of any
+     * un-deserializable ("invalid") log events within the range
+     * @return null if any log event in the range does not exist (e.g., the range exceeds the number
+     * of log events in the file).
+     */
+    [[nodiscard]] auto build_idx(size_t begin_idx, size_t end_idx) -> emscripten::val;
+
+    /**
+     * Decodes the deserialized log events in the range `[beginIdx, endIdx)`.
+     *
+     * @param beginIdx
+     * @param endIdx
+     * @return The decoded log events on success.
+     * @return null if any log event in the range doesn't exist (e.g., the range exceeds the number
+     * of log events in the file).
+     */
+    [[nodiscard]] auto decode(size_t begin_idx, size_t end_idx) const -> emscripten::val;
+
+private:
+    // Constructor
+    explicit ClpIrV1Decoder(
+            std::unique_ptr<char const[]>&& data_buffer,
+            std::shared_ptr<clp::streaming_compression::zstd::Decompressor> zstd_decompressor,
+            clp::ir::LogEventDeserializer<clp::ir::four_byte_encoded_variable_t> deserializer
+    );
+
+    // Variables
+    bool m_full_range_built{false};
+    std::vector<clp::ir::LogEvent<clp::ir::four_byte_encoded_variable_t>> m_log_events;
```

Comment:
How about `m_encoded_log_events`?

Reply:
I'm not too familiar with the naming conventions in the encoders/decoders but I believe differences here are a matter of perspective. Would it be more appropriate to name it `m_deserialized_log_events`? Or do you mean the messages in the buffered log events objects are still being "encoded"? (then it makes sense for us to "decode" those later)

For now, I have renamed it as `m_encoded_log_events`.

Reply:
Yeah, it is a matter of perspective. I prefer `m_encoded_log_events` since it indicates the current state of the log events where `m_deserialized_log_events` indicates the state change that's occurred. Do you think that makes sense?